contracts/contracts/harvest/Harvester.sol

```diff
@@ -0,0 +1,388 @@
+            }
+        }
+
+        emit RewardTokenConfigUpdated(
+            _tokenAddress,
+            _allowedSlippageBps,
+            _harvestRewardBps,
+            _uniswapV2CompatibleAddr,
+            _liquidationLimit
+        );
+    }
+
+    /***************************************
+                    Rewards
+    ****************************************/
+
+    /**
+     * @dev Transfer token to governor. Intended for recovering tokens stuck in
+     *      contract, i.e. mistaken sends.
+     * @param _asset Address for the asset
+     * @param _amount Amount of the asset to transfer
+     */
+    function transferToken(address _asset, uint256 _amount)
+        external
+        onlyGovernor
+    {
+        IERC20(_asset).safeTransfer(governor(), _amount);
+    }
+
+    /**
+     * @dev Collect reward tokens from all strategies
+     */
+    function harvest() external onlyGovernor nonReentrant {
+        _harvest();
+    }
+
+    /**
+     * @dev Swap all supported swap tokens for stablecoins via Uniswap.
+     */
+    function swap() external onlyGovernor nonReentrant {
+        _swap();
+    }
+
+    /*
+     * @dev Collect reward tokens from all strategies and swap for supported
+     *      stablecoin via Uniswap
+     */
+    function harvestAndSwap() external nonReentrant {
+        // TODO: add protection so that harvestAndSwap isn't called twice too closely together
+        _harvest();
+        _swap();
+        if (msg.sender != vaultAddress) {
+            IVault(vaultAddress).rebase();
+        }
+    }
+
+    /**
+     * @dev Collect reward tokens from all strategies
+     */
+    function _harvest() internal {
+        address[] memory allStrategies = IVault(vaultAddress)
+            .getAllStrategies();
+        for (uint256 i = 0; i < allStrategies.length; i++) {
+            _harvest(allStrategies[i]);
+        }
+    }
+
+    /**
+     * @dev Collect reward tokens for a specific strategy. Called from the vault.
+     * @param _strategyAddr Address of the strategy to collect rewards from
+     */
+    function harvest(address _strategyAddr) external onlyGovernor nonReentrant {
+        // TODO: add protection so that harvestAndSwap isn't called twice too closely together
+        _harvest(_strategyAddr);
+    }
+
+    /**
+     * @dev Collect reward tokens for a specific strategy and swap for supported
+     *      stablecoin via Uniswap. Called from the vault.
+     * @param _strategyAddr Address of the strategy to collect rewards from
+     */
+    function harvestAndSwap(address _strategyAddr) external nonReentrant {
+        IStrategy strategy = IStrategy(_strategyAddr);
+        _harvest(address(strategy));
+        address[] memory rewardTokens = strategy.getRewardTokenAddresses();
+        for (uint256 i = 0; i < rewardTokens.length; i++) {
+            _swap(rewardTokens[i]);
+        }
+        if (msg.sender != vaultAddress) {
```

Comment:
@DanielVF @tomlinton made this change and an identical one in a `harvestAndSwap()` without strategy address parameter. The issues is that Vault's `_allocate` calls this function and if we call back `rebase` the re-entrancy gets triggered. Also we probably do not want Vault's `allocate` to also trigger a rebase.

Reply:
Vault allocate should no longer harvest - just move funds into strategies. Sorry, I had not gotten to reviewing the vault changes yet.

Reply:
Just pasting it here for easier reference. This was what Vault used to do before allocating: https://github.com/OriginProtocol/origin-dollar/blob/master/contracts/contracts/vault/VaultCore.sol#L218-L241

And that has been shortened to this: https://github.com/OriginProtocol/origin-dollar/pull/869/files#diff-113977321a6b18ff2702753444615646663eef373deadf81107bf316f1300448R217-R219

So to double confirm we are ok with removing the harvesting bit from `_allocate` call? It makes sense to me, and it should make allocates a lot more gas efficient.

Reply:
Yes. I am good removing the harvest from the vault.

Reply:
👍

Reply:
Done